File: 02.05/baekjoon_7453_박동현.py
'''
합이 0인 네 정수

MTIM, 투 포인터

완전 동일한 로직으로,

python 시간 초과
pypy 689172kb, 6968ms
go 255836kb, 7528ms

go가 지는 경우 처음봄; 근데 go가 메모리 효율은 압도적.
'''
import sys; input = sys.stdin.readline


def find(arr, idx):
    target = arr[idx]
    res = 0
    for i in range(idx, N*N):
        if target == arr[i]:
            res += 1
        else: break
    return res

N = int(input())
A,B,C,D = [], [], [], []

for _ in range(N):
    a,b,c,d = map(int,input().split())
    A.append(a)
    B.append(b)
    C.append(c)
    D.append(d)

AB, CD = [], []
for i in range(N):
    for j in range(N):
        AB.append(A[i] + B[j])
        CD.append(C[i] + D[j])

AB.sort(); CD.sort(reverse=True)

ab, cd = 0, 0
ans = 0
while cd < N*N and ab < N*N:
    total = AB[ab] + CD[cd]
    if total == 0:
        ab_cnt = find(AB, ab)
        cd_cnt = find(CD, cd)
        ans += ab_cnt * cd_cnt
        ab += ab_cnt
        cd += cd_cnt
    elif total > 0: cd += find(CD, cd)
    else: ab += find(AB, ab)

print(ans)

File: 25/2/02.05/baekjoon_7453_박동현.go
package main

import (
	"bufio"
	"os"
	"sort"
	"strconv"
)

var (
	bs = bufio.NewScanner(os.Stdin)
	bw = bufio.NewWriter(os.Stdout)
)

func find(arr []int, idx int) int {
	target := arr[idx]
	res := 0
	for i := idx; i < len(arr); i++ {
		if target == arr[i] {
			res++
		} else {
			break
		}
	}
	return res
}

func main() {
	bs.Split(bufio.ScanWords)
	defer bw.Flush()

	N := input()
	A, B, C, D := make([]int, N), make([]int, N), make([]int, N), make([]int, N)
	for i := 0; i < N; i++ {
		A[i] = input()
		B[i] = input()
		C[i] = input()
		D[i] = input()
	}
	AB := make([]int, N*N)
	CD := make([]int, N*N)
	for i := 0; i < N; i++ {
		for j := 0; j < N; j++ {
			AB[i*N+j] = A[i] + B[j]
			CD[i*N+j] = C[i] + D[j]
		}
	}
	sort.Ints(AB)
	sort.Sort(sort.Reverse(sort.IntSlice(CD)))

	ab, cd := 0, 0
	ans := 0
	for ab < N*N && cd < N*N {
		total := AB[ab] + CD[cd]
		if total == 0 {
			abCnt := find(AB, ab)
			cdCnt := find(CD, cd)
			ans += abCnt * cdCnt
			ab += abCnt
			cd += cdCnt
		} else if total > 0 {
			cd += find(CD, cd)
		} else {
			ab += find(AB, ab)
		}
	}
	bw.WriteString(strconv.Itoa(ans))
}

func input() int {
	bs.Scan()
	num, _ := strconv.Atoi(bs.Text())
	return num
}
